Part A. Baseline Model

In [1]:
#Baseline Model
import numpy as np
import pandas as pd
from sklearn import metrics

concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/sklearn/utils/validation.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  LARGE_SPARSE_SUPPORTED = LooseVersion(scipy_version) >= '0.14.0'


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [2]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
target = concrete_data['Strength']

In [3]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [4]:
#Normalized Predictors
predictors_norm = (predictors - predictors.mean())/predictors.std()

In [5]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.3)
print('Train: ', X_train.shape)
print('Test: ', X_test.shape)

Train:  (721, 8)
Test:  (309, 8)


In [6]:
#Train and Test Network

def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [7]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
n_cols = predictors_norm.shape[1]

In [8]:
model = regression_model()

In [9]:
model.fit(X_train, y_train, epochs = 100, verbose = 2)



Epoch 1/100


2023-07-22 05:20:39.128731: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2023-07-22 05:20:39.133332: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2394330000 Hz
2023-07-22 05:20:39.134212: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x561c1dc5ced0 executing computations on platform Host. Devices:
2023-07-22 05:20:39.134270: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2023-07-22 05:20:39.211864: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line fl

 - 0s - loss: 1545.0881
Epoch 2/100
 - 0s - loss: 1525.6669
Epoch 3/100
 - 0s - loss: 1506.0333
Epoch 4/100
 - 0s - loss: 1485.8066
Epoch 5/100
 - 0s - loss: 1465.4628
Epoch 6/100
 - 0s - loss: 1444.1320
Epoch 7/100
 - 0s - loss: 1422.0450
Epoch 8/100
 - 0s - loss: 1399.3694
Epoch 9/100
 - 0s - loss: 1375.6746
Epoch 10/100
 - 0s - loss: 1351.2572
Epoch 11/100
 - 0s - loss: 1326.2562
Epoch 12/100
 - 0s - loss: 1300.1428
Epoch 13/100
 - 0s - loss: 1273.7037
Epoch 14/100
 - 0s - loss: 1246.5237
Epoch 15/100
 - 0s - loss: 1218.3087
Epoch 16/100
 - 0s - loss: 1190.1992
Epoch 17/100
 - 0s - loss: 1160.4869
Epoch 18/100
 - 0s - loss: 1130.8402
Epoch 19/100
 - 0s - loss: 1100.7989
Epoch 20/100
 - 0s - loss: 1069.9684
Epoch 21/100
 - 0s - loss: 1039.1987
Epoch 22/100
 - 0s - loss: 1007.5201
Epoch 23/100
 - 0s - loss: 976.4715
Epoch 24/100
 - 0s - loss: 944.7441
Epoch 25/100
 - 0s - loss: 912.9793
Epoch 26/100
 - 0s - loss: 882.1756
Epoch 27/100
 - 0s - loss: 850.9291
Epoch 28/100
 - 0s - loss: 

In [10]:
pred = model.predict(X_test)
mse = metrics.mean_squared_error(y_test, pred)
print('MSE Norm: ', mse)

MSE Norm:  162.78431952927292


In [11]:
array_MSE = []

for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.3)
    model = regression_model()
    model.fit(X_train, y_train, epochs = 100, verbose = 2)
    y_pred = model.predict(X_test)
    array_MSE.append(metrics.mean_squared_error(y_test, y_pred))

Epoch 1/100
 - 0s - loss: 1517.6616
Epoch 2/100
 - 0s - loss: 1500.2815
Epoch 3/100
 - 0s - loss: 1482.2490
Epoch 4/100
 - 0s - loss: 1463.7295
Epoch 5/100
 - 0s - loss: 1444.5309
Epoch 6/100
 - 0s - loss: 1424.1580
Epoch 7/100
 - 0s - loss: 1403.0337
Epoch 8/100
 - 0s - loss: 1380.9551
Epoch 9/100
 - 0s - loss: 1357.6584
Epoch 10/100
 - 0s - loss: 1333.4508
Epoch 11/100
 - 0s - loss: 1307.7431
Epoch 12/100
 - 0s - loss: 1281.5487
Epoch 13/100
 - 0s - loss: 1253.7554
Epoch 14/100
 - 0s - loss: 1225.1795
Epoch 15/100
 - 0s - loss: 1195.4813
Epoch 16/100
 - 0s - loss: 1164.4800
Epoch 17/100
 - 0s - loss: 1133.1034
Epoch 18/100
 - 0s - loss: 1100.6356
Epoch 19/100
 - 0s - loss: 1067.5012
Epoch 20/100
 - 0s - loss: 1033.6440
Epoch 21/100
 - 0s - loss: 999.3413
Epoch 22/100
 - 0s - loss: 965.1649
Epoch 23/100
 - 0s - loss: 930.0564
Epoch 24/100
 - 0s - loss: 895.5288
Epoch 25/100
 - 0s - loss: 861.3245
Epoch 26/100
 - 0s - loss: 826.4727
Epoch 27/100
 - 0s - loss: 792.1921
Epoch 28/100
 - 0

In [12]:
#Mean and Std Dev
print("Mean MSE: ", np.mean(array_MSE))
print("Standard Deviation MSE: ", np.std(array_MSE))

Mean MSE:  159.25541097525567
Standard Deviation MSE:  17.75368671106919


How does the mean of the MSE compare to that from Step B?

Unliked Part B, the mean of the MSE seems to have decreased significantly from 373.21 to 159.25. Similarly, the standard deviation decreased from 78.21 to 17.75.

375.86 to 373.21, which is insignificant. On the other hand, the standard deviation decreased heavily with the normalization from 451.35 to 78.21.
